infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/AsyncStoragePodWatcher.java

```diff
@@ -0,0 +1,171 @@
+  }
+
+  /**
+   * Checking current system configuration on ability to run Async Storage Pod. Will be checked next
+   * value of properties:
+   *
+   * <ul>
+   *   <li>che.infra.kubernetes.namespace.default=<username>-che
+   *   <li>che.infra.kubernetes.namespace.allow_user_defined=false
+   *   <li>che.infra.kubernetes.pvc.strategy=common
+   *   <li>che.limits.user.workspaces.run.count=1
+   * </ul>
+   */
+  private boolean isAsyncStoragePodCanBeRun(
+      String pvcStrategy,
+      boolean allowUserDefinedNamespaces,
+      String defaultNamespaceName,
+      int runtimesPerUser) {
+    return !allowUserDefinedNamespaces
+        && COMMON_STRATEGY.equals(pvcStrategy)
+        && runtimesPerUser == 1
+        && !isNullOrEmpty(defaultNamespaceName)
+        && defaultNamespaceName.contains("<username>");
+  }
+
+  @ScheduleDelay(
+      unit = TimeUnit.MINUTES,
+      initialDelay = 1,
+      delayParameterName = "che.infra.kubernetes.async.storage.shutdown_check_period_min")
+  public void check() {
+    if (isAsyncStoragePodCanBeRun) { // if system not support async storage mode do nothing
+      for (User user :
+          iterateLazily((maxItems, skipCount) -> userManager.getAll(maxItems, skipCount))) {
+        try {
+          String owner = user.getId();
+          Map<String, String> preferences = preferenceManager.find(owner);
+          String lastTimeAccess = preferences.get(LAST_ACTIVITY_TIME);
+          String namespace = preferences.get(LAST_ACTIVE_INFRASTRUCTURE_NAMESPACE);
+          if (isNullOrEmpty(namespace)
+              || isNullOrEmpty(lastTimeAccess)
+              || isAnyRuntimeInProgress(owner)) {
+            continue;
+          }
+          long lastTimeAccessSec = parseLong(lastTimeAccess);
+          long epochSec = clock.instant().getEpochSecond();
+          if (epochSec - lastTimeAccessSec >= shutdownTimeoutSec) {
+            PodResource<Pod, DoneablePod> podDoneablePodPodResource =
+                kubernetesClientFactory
+                    .create()
+                    .pods()
+                    .inNamespace(namespace)
+                    .withName(ASYNC_STORAGE);
+            if (podDoneablePodPodResource.get() != null) {
+              podDoneablePodPodResource.delete();
+            }
+          }
+        } catch (InfrastructureException | ServerException e) {
+          LOG.error(e.getMessage(), e);
+        }
+      }
+    }
+  }
+
+  /**
+   * Going to check is currently owner has workspaces in progress: it's status is {@link
+   * org.eclipse.che.api.core.model.workspace.WorkspaceStatus#STARTING} or {@link
+   * org.eclipse.che.api.core.model.workspace.WorkspaceStatus#STOPPING})
+   *
+   * @param owner the user id to check
+   */
+  private boolean isAnyRuntimeInProgress(String owner)
+      throws ServerException, InfrastructureException {
+    Set<String> inProgress = runtimes.getInProgress();
```

Comment:
can we get this before user's loop, so we're not calling it multiple times for same result?